File: benchmark_release/public/case_000008/src/app/services/notifications.py
"""
Notification service — creates in-app notifications for users.
"""
from datetime import datetime


def notify(conn, user_id, message):
    now = datetime.utcnow().isoformat()
    conn.execute(
        'INSERT INTO notifications (user_id, message, read, created_at) VALUES (?,?,?,?)',
        (user_id, message, 0, now)
    )
    conn.commit()


def get_unread(conn, user_id):
    rows = conn.execute(
        '''SELECT * FROM notifications
           WHERE user_id=? AND read=0
           ORDER BY created_at DESC''',
        (user_id,)
    ).fetchall()
    return [dict(r) for r in rows]


def mark_read(conn, user_id, notification_id):
    conn.execute(
        'UPDATE notifications SET read=1 WHERE id=? AND user_id=?',
        (notification_id, user_id)
    )
    conn.commit()

File: benchmark_controls_release/public/case_000012_clean/src/app/services/notifications.js
'use strict';

const db = require('../models/db');

function listForUser(userId) {
  return db.prepare(
    'SELECT * FROM notifications WHERE user_id = ? ORDER BY id DESC LIMIT 30'
  ).all(userId);
}

function send(userId, message) {
  db.prepare('INSERT INTO notifications (user_id, message, read) VALUES (?, ?, 0)').run(userId, message);
}

function markRead(id, userId) {
  db.prepare('UPDATE notifications SET read = 1 WHERE id = ? AND user_id = ?').run(id, userId);
}

function markAllRead(userId) {
  db.prepare('UPDATE notifications SET read = 1 WHERE user_id = ?').run(userId);
}

function unreadCount(userId) {
  return db.prepare('SELECT COUNT(*) AS cnt FROM notifications WHERE user_id = ? AND read = 0').get(userId).cnt;
}

module.exports = { listForUser, send, markRead, markAllRead, unreadCount };